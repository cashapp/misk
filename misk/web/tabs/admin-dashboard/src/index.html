<!DOCTYPE html>
<html>

<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
  <meta http-equiv="X-UA-Compatible" content="ie=edge" />
  <link rel="shortcut icon" type="image/x-icon" href="/static/favicon.ico" />
</head>

<body>
  <div id="admin-dashboard"></div>

  <!-- Misk Libraries -->
  <link rel="stylesheet" type="text/css" href="/@misk/common/styles.css" />
  <script type="text/javascript" src="/@misk/common/vendors.js" preload></script>
  <script type="text/javascript" src="/@misk/common/common.js" preload></script>
  <script type="text/javascript" src="/@misk/core/core.js" preload></script>
  <script type="text/javascript" src="/@misk/simpleredux/simpleredux.js" preload></script>

  <!-- Tab Admin Dashboard -->
  <script type="text/javascript" src="/_admin/tab_admin-dashboard.js"></script>
</body>

</html>
