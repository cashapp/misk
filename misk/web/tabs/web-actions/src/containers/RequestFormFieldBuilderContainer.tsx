/** @jsx jsx */
import {
  Button,
  ButtonGroup,
  Card,
  ControlGroup,
  FormGroup,
  HTMLSelect,
  InputGroup,
  Intent,
  Label,
  TextArea,
  Tooltip
} from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import { css, jsx } from "@emotion/core"
import { WrapTextContainer } from "@misk/core"
import {
  onChangeFnCall,
  onChangeToggleFnCall,
  onClickFnCall,
  simpleSelectorGet
} from "@misk/simpleredux"
import { OrderedMap } from "immutable"
import { connect } from "react-redux"
import {
  cssButton,
  cssTooltip,
  cssWrapTextArea,
  Metadata,
  QuantityButton
} from "../components"
import {
  BaseFieldTypes,
  IDispatchProps,
  IState,
  ITypesFieldMetadata,
  IWebActionInternal,
  mapDispatchToProps,
  mapStateToProps,
  padId,
  parseEnumType,
  TypescriptBaseTypes,
  ServerTypes
} from "../ducks"

const cssCard = css`
  box-shadow: none !important;
  border-left: 2px black;
  margin-bottom: 0px;
  padding: 0px 0px 0px 15px !important;
`

const cssFormGroup = css`
  margin: 0px !important;
`

const repeatableFieldButtons = (
  props: {
    action: IWebActionInternal
    id: string
    tag: string
    typesMetadata: OrderedMap<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => {
  const { action, id, tag, typesMetadata } = props
  const metadata = typesMetadata.get(id)
  const editRawButton = (
    <Tooltip
      css={css(cssTooltip)}
      content={"Edit raw input"}
      key={`editRaw${id}`}
      lazy={true}
    >
      <Button
        active={simpleSelectorGet(
          props.simpleRedux,
          [`${tag}::EditRawButton:${padId(id)}`, "data"],
          false
        )}
        css={css(cssButton)}
        defaultValue={simpleSelectorGet(props.simpleRedux, [
          `${tag}::${padId(id)}`,
          "data"
        ])}
        icon={IconNames.MORE}
        onClick={onChangeToggleFnCall(
          props.simpleMergeToggle,
          `${tag}::EditRawButton:${padId(id)}`,
          props.simpleRedux
        )}
      />
    </Tooltip>
  )
  const dirtyInput = metadata.dirtyInput
  const dirtyInputButton = (
    <Tooltip
      css={css(cssTooltip)}
      content={dirtyInput ? "Remove from request body" : "Add to request body"}
      key={`dirtyInput${id}`}
      lazy={true}
    >
      <Button
        icon={dirtyInput ? IconNames.REMOVE : IconNames.ADD_TO_ARTIFACT}
        onClick={(event: any) => {
          props.simpleMergeData(`${tag}::ButtonRequestBody`, true)
          ;(dirtyInput
            ? props.webActionsUnsetDirtyInput
            : props.webActionsSetDirtyInput)(id, action, props.webActionsRaw)
        }}
      />
    </Tooltip>
  )
  if (
    metadata &&
    metadata.id !== "0" &&
    typesMetadata.get(metadata.idParent).repeated
  ) {
    const { idParent } = metadata
    const { idChildren } = typesMetadata.get(idParent)
    const addButton = (
      <QuantityButton
        action={action}
        changeFn={props.webActionsAdd}
        content={"Add Field"}
        icon={IconNames.PLUS}
        id={idParent}
        key={`Add${id}`}
        oldState={props.webActionsRaw}
      />
    )
    const removeButton = idChildren.size > 1 && (
      <QuantityButton
        action={action}
        changeFn={props.webActionsRemove}
        content={"Remove Field"}
        icon={IconNames.CROSS}
        id={id}
        key={`Remove${id}`}
        oldState={props.webActionsRaw}
      />
    )
    return [dirtyInputButton, editRawButton, addButton, removeButton]
  } else if (metadata && !metadata.repeated && metadata.typescriptType) {
    return [dirtyInputButton, editRawButton]
  } else if (metadata && id !== "0") {
    return [dirtyInputButton]
  } else if (metadata && metadata.id === "0") {
    return []
  } else {
    return [<Button css={css(cssButton)} icon={IconNames.WARNING_SIGN} />]
  }
}

const EditRawInput = (
  props: { children: any; id: string; tag: string } & IDispatchProps & IState
) => {
  const { children, id, tag } = props
  const editRawIsOpen = simpleSelectorGet(
    props.simpleRedux,
    [`${tag}::EditRawButton:${padId(id)}`, "data"],
    false
  )
  if (editRawIsOpen) {
    return (
      <TextArea
        css={css(cssWrapTextArea)}
        defaultValue={simpleSelectorGet(props.simpleRedux, [
          `${tag}::${padId(id)}`,
          "data"
        ])}
        fill={true}
        growVertically={true}
        onBlur={onChangeFnCall(props.simpleMergeData, `${tag}::${padId(id)}`)}
      />
    )
  } else {
    return children
  }
}

const formLabelFormatter = (name: string, serverType: string) => {
  if (name && serverType && serverType.startsWith(ServerTypes.Enum)) {
    const { enumClassName } = parseEnumType(serverType)
    return <WrapTextContainer>{`${name} (${enumClassName})`}</WrapTextContainer>
  } else if (name && serverType) {
    return <WrapTextContainer>{`${name} (${serverType})`}</WrapTextContainer>
  } else {
    return <span />
  }
}

const clickDirtyInputFns = (
  props: {
    action: IWebActionInternal
    id: string
    tag: string
    typesMetadata: OrderedMap<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => () => {
  const { action, id, tag } = props
  props.webActionsSetDirtyInput(id, action, props.webActionsRaw)
  props.simpleMergeData(`${tag}::ButtonRequestBody`, true)
}

const UnconnectedRequestFormFieldBuilderContainer = (
  props: {
    action: IWebActionInternal
    id: string
    tag: string
    typesMetadata: OrderedMap<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => {
  const { action, id, tag, typesMetadata } = props
  const metadata = typesMetadata.get(id)
  if (metadata) {
    const {
      name,
      idChildren,
      serverType,
      typescriptType
    } = metadata as ITypesFieldMetadata
    if (typescriptType === TypescriptBaseTypes.boolean) {
      return (
        <FormGroup
          css={css(cssFormGroup)}
          label={formLabelFormatter(name, serverType)}
        >
          <ControlGroup>
            {...repeatableFieldButtons({ ...props, id })}
            <EditRawInput {...props} id={id} tag={tag}>
              <Button
                css={css(cssButton)}
                defaultValue={simpleSelectorGet(props.simpleRedux, [
                  `${tag}::${padId(id)}`,
                  "data"
                ])}
                intent={
                  simpleSelectorGet(
                    props.simpleRedux,
                    [`${tag}::${padId(id)}`, "data"],
                    false
                  )
                    ? Intent.PRIMARY
                    : Intent.WARNING
                }
                onChange={onChangeFnCall(clickDirtyInputFns(props))}
                onClick={() => {
                  props.simpleMergeToggle(
                    `${tag}::${padId(id)}`,
                    props.simpleRedux
                  )
                  clickDirtyInputFns(props)()
                }}
              >
                {simpleSelectorGet(props.simpleRedux, [
                  `${tag}::${padId(id)}`,
                  "data"
                ]).toString() || "unset"}
              </Button>
            </EditRawInput>
          </ControlGroup>
        </FormGroup>
      )
    } else if (typescriptType === TypescriptBaseTypes.number) {
      return (
        <FormGroup
          css={css(cssFormGroup)}
          label={formLabelFormatter(name, serverType)}
        >
          <ControlGroup>
            {...repeatableFieldButtons({ ...props, id })}
            <EditRawInput {...props} id={id} tag={tag}>
              <InputGroup
                defaultValue={simpleSelectorGet(props.simpleRedux, [
                  `${tag}::${padId(id)}`,
                  "data"
                ])}
                onChange={onChangeFnCall(clickDirtyInputFns(props))}
                onClick={onChangeFnCall(clickDirtyInputFns(props))}
                onBlur={onChangeFnCall(
                  props.simpleMergeData,
                  `${tag}::${padId(id)}`
                )}
                placeholder={serverType}
              />
            </EditRawInput>
          </ControlGroup>
        </FormGroup>
      )
    } else if (typescriptType === TypescriptBaseTypes.string) {
      return (
        <FormGroup
          css={css(cssFormGroup)}
          label={formLabelFormatter(name, serverType)}
        >
          <ControlGroup>
            {...repeatableFieldButtons({ ...props, id })}
            <EditRawInput {...props} id={id} tag={tag}>
              <InputGroup
                defaultValue={simpleSelectorGet(props.simpleRedux, [
                  `${tag}::${padId(id)}`,
                  "data"
                ])}
                onBlur={onChangeFnCall(
                  props.simpleMergeData,
                  `${tag}::${padId(id)}`
                )}
                onChange={onChangeFnCall(clickDirtyInputFns(props))}
                onClick={onChangeFnCall(clickDirtyInputFns(props))}
                placeholder={serverType}
              />
            </EditRawInput>
          </ControlGroup>
        </FormGroup>
      )
    } else if (typescriptType === TypescriptBaseTypes.enum) {
      const { enumValues } = parseEnumType(serverType)
      return (
        <FormGroup
          css={css(cssFormGroup)}
          label={formLabelFormatter(name, serverType)}
        >
          <ControlGroup>
            {...repeatableFieldButtons({ ...props, id })}
            <EditRawInput {...props} id={id} tag={tag}>
              <HTMLSelect
                large={true}
                onChange={onChangeFnCall(
                  props.simpleMergeData,
                  `${tag}::${padId(id)}`
                )}
                onClick={onChangeFnCall(clickDirtyInputFns(props))}
                onBlur={onChangeFnCall(
                  props.simpleMergeData,
                  `${tag}::${padId(id)}`
                )}
                // Show empty option to prompt selection since first option is not automatically persisted
                options={[""].concat(enumValues)}
              />
            </EditRawInput>
          </ControlGroup>
        </FormGroup>
      )
    } else if (typescriptType === null && idChildren.size > 0) {
      if (
        idChildren.first() &&
        typesMetadata.get(idChildren.first() as string).typescriptType &&
        BaseFieldTypes.hasOwnProperty(serverType)
      ) {
        return (
          <div>
            {idChildren.map((child: string) => (
              <RequestFormFieldBuilderContainer
                action={action}
                id={child}
                key={child}
                tag={tag}
                typesMetadata={typesMetadata}
              />
            ))}
          </div>
        )
      } else {
        const fieldGroup = (child: string, parentServerType: string) => {
          const {
            idChildren: grandChildren,
            name: childName,
            serverType: childServerType
          } = typesMetadata.get(child)
          const parentOfLeaf =
            grandChildren.reduce(
              (sum: number, grandchild: string) =>
                typesMetadata.get(grandchild).idChildren.size + sum,
              0
            ) === 0
          const noParentGroupLabel = parentOfLeaf && parentServerType === null
          if (
            BaseFieldTypes.hasOwnProperty(childServerType) ||
            noParentGroupLabel
          ) {
            return (
              <div key={child}>
                <RequestFormFieldBuilderContainer
                  action={action}
                  id={child}
                  tag={tag}
                  typesMetadata={typesMetadata}
                />
              </div>
            )
          } else {
            return (
              <div key={child}>
                <Label>{formLabelFormatter(childName, childServerType)}</Label>
                <ButtonGroup>
                  {...repeatableFieldButtons({ ...props, id: child })}
                </ButtonGroup>
                <RequestFormFieldBuilderContainer
                  action={action}
                  id={child}
                  tag={tag}
                  typesMetadata={typesMetadata}
                />
              </div>
            )
          }
        }
        if (id === "0") {
          const whichFormData = simpleSelectorGet(
            props.simpleRedux,
            [`${tag}::RequestBodyFormInputType`, "data"],
            false
          )
            ? "RAW"
            : "FORM"
          return (
            <div>
              <ControlGroup fill={true}>
                <Button
                  active={whichFormData === "FORM"}
                  css={css(cssButton)}
                  icon={IconNames.FORM}
                  onClick={onChangeToggleFnCall(
                    props.simpleMergeToggle,
                    `${tag}::RequestBodyFormInputType`,
                    props.simpleRedux
                  )}
                  text={"Form"}
                />
                <Button
                  active={whichFormData === "RAW"}
                  css={css(cssButton)}
                  icon={IconNames.MORE}
                  onClick={onChangeToggleFnCall(
                    props.simpleMergeToggle,
                    `${tag}::RequestBodyFormInputType`,
                    props.simpleRedux
                  )}
                  text={"Raw"}
                />
              </ControlGroup>
              {whichFormData === "FORM" ? (
                <Card css={css(cssCard)}>
                  {idChildren.map((child: string) =>
                    fieldGroup(child, serverType)
                  )}
                </Card>
              ) : (
                <Card css={css(cssCard)}>
                  <TextArea
                    css={css(cssWrapTextArea)}
                    defaultValue={simpleSelectorGet(props.simpleRedux, [
                      `${tag}::RawRequestBody`,
                      "data"
                    ])}
                    fill={true}
                    growVertically={true}
                    onChange={onChangeFnCall(
                      props.simpleMergeData,
                      `${tag}::RawRequestBody`
                    )}
                    placeholder={
                      "Raw request body. This input will return a string or JSON."
                    }
                  />
                </Card>
              )}
            </div>
          )
        } else {
          return (
            <Card css={css(cssCard)}>
              {idChildren.map((child: string) => fieldGroup(child, serverType))}
            </Card>
          )
        }
      }
    } else {
      return (
        <FormGroup
          css={css(cssFormGroup)}
          label={formLabelFormatter(name, serverType)}
        >
          <ControlGroup>
            {...repeatableFieldButtons({ ...props, id })}
            <TextArea
              css={css(cssWrapTextArea)}
              fill={true}
              growVertically={true}
              onBlur={(event: any) => {
                props.simpleMergeData(
                  `${tag}::${padId(id)}`,
                  event.target.value
                )
                clickDirtyInputFns(props)()
              }}
              onChange={onClickFnCall(clickDirtyInputFns(props))}
              onClick={onClickFnCall(clickDirtyInputFns(props))}
              placeholder={
                "Unparseable type. This input will return a string or JSON."
              }
            />
          </ControlGroup>
        </FormGroup>
      )
    }
  } else {
    return (
      <Metadata
        content={`No Request Body for ${action.name} ${action.pathPattern}`}
      />
    )
  }
}

const RequestFormFieldBuilderContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(UnconnectedRequestFormFieldBuilderContainer)

export default RequestFormFieldBuilderContainer
