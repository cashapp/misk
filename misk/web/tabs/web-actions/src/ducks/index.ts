import {
  dispatchSimpleRedux,
  IDispatchSimpleRedux,
  ISimpleReduxImmutableState,
  ISimpleReduxState,
  SimpleReduxReducer,
  SimpleReduxSaga,
  simpleRootSelector,
  watchSimpleReduxSagas
} from "@misk/simpleredux"
import {
  connectRouter,
  LocationChangeAction,
  RouterState
} from "connected-react-router"
import { History } from "history"
import { AnyAction, combineReducers, Reducer } from "redux"
import { all, fork } from "redux-saga/effects"
import {
  dispatchWebActions,
  IDispatchWebActions,
  IWebActionsImmutableState,
  IWebActionsState,
  watchWebActionsSagas,
  WebActionsReducer
} from "./webActions"
export * from "./webActions"

/**
 * Redux Store State
 */
export interface IState {
  router: Reducer<RouterState, LocationChangeAction>
  simpleRedux: ISimpleReduxState
  webActions: IWebActionsState
  webActionsRaw: IWebActionsImmutableState
}

/**
 * Dispatcher
 */
export interface IDispatchProps
  extends IDispatchSimpleRedux,
    IDispatchWebActions {}

export const rootDispatcher: IDispatchProps = {
  ...dispatchSimpleRedux,
  ...dispatchWebActions
}

/**
 * State Selectors
 */
export const rootSelectors = (state: IState) => ({
  router: state.router,
  simpleRedux: simpleRootSelector<IState, ISimpleReduxImmutableState>(
    "simpleRedux",
    state
  ),
  webActions: simpleRootSelector<IState, IWebActionsImmutableState>(
    "webActions",
    state
  ),
  webActionsRaw: state.webActionsRaw
})

/**
 * Reducers
 */
export const rootReducer = (history: History): Reducer<any, AnyAction> =>
  combineReducers({
    router: connectRouter(history),
    simpleRedux: SimpleReduxReducer,
    webActions: WebActionsReducer,
    webActionsRaw: WebActionsReducer
  })

/**
 * Sagas
 */
export function* rootSaga(): SimpleReduxSaga {
  yield all([fork(watchWebActionsSagas), fork(watchSimpleReduxSagas)])
}

/**
 * Map Dispatch/State to Props
 */
export const mapStateToProps = (state: IState) => rootSelectors(state)

export const mapDispatchToProps: IDispatchProps = {
  ...rootDispatcher
}
