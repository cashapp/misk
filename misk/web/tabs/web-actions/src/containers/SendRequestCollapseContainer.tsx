/** @jsx jsx */
import {
  Button,
  Collapse,
  ControlGroup,
  HTMLSelect,
  InputGroup,
  Menu
} from "@blueprintjs/core"
import { jsx } from "@emotion/core"
import { CodePreContainer, FlexContainer, HTTPMethodIntent } from "@misk/core"
import {
  HTTPMethodDispatch,
  onChangeFnCall,
  simpleSelectorGet
} from "@misk/simpleredux"
import { HTTPMethod } from "http-method-enum"
import { connect } from "react-redux"
import {
  cssButton,
  cssColumn,
  cssMetadataMenu,
  Metadata,
  MetadataCollapse,
  MetadataCopyToClipboard,
  StatusTagComponent
} from "../components"
import { RequestFormContainer } from "../containers"
import {
  getFormData,
  IDispatchProps,
  IState,
  IWebActionInternal,
  mapDispatchToProps,
  mapStateToProps,
  methodHasBody
} from "../ducks"

const RequestBodyForm = (
  props: {
    action: IWebActionInternal
    method: HTTPMethod
    tag: string
  } & IState &
    IDispatchProps
) => {
  const { method, tag } = props
  if (methodHasBody(props.method)) {
    return (
      <MetadataCollapse
        label={"Input"}
        tag={`${tag}::ButtonFormRequestBody`}
        text={"Request Body"}
      >
        <RequestFormContainer {...props} tag={tag} />
      </MetadataCollapse>
    )
  } else {
    return (
      <Metadata
        {...props}
        content={"Request Body"}
        label={`${method} does not have a Body`}
      />
    )
  }
}

/**
 * Collapse wrapped Send a Request form for each Web Action card
 */
const SendRequestCollapseContainer = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { action, tag } = props

  // Determine if Send Request form for the Web Action should be open
  const isOpen = simpleSelectorGet(props.simpleRedux, [
    `${tag}::ButtonSendRequest`,
    "data"
  ])
  const url = simpleSelectorGet(props.simpleRedux, [`${tag}::URL`, "data"])

  // Pre-populate the URL field with the action path pattern on open of request form
  if (isOpen && !url) {
    props.simpleMergeData(`${tag}::URL`, action.pathPattern)
  }
  const method: HTTPMethod =
    simpleSelectorGet(props.simpleRedux, [`${tag}::Method`, "data"]) ||
    action.httpMethod.reverse()[0]

  // Response with fallback to error
  const response = simpleSelectorGet(
    props.simpleRedux,
    [`${tag}::Response`, "data"],
    simpleSelectorGet(props.simpleRedux, [`${tag}::Response`, "error"])
  )

  // Response.data with fallback to error.response
  const responseData = simpleSelectorGet(
    props.simpleRedux,
    [`${tag}::Response`, "data", "data"],
    simpleSelectorGet(props.simpleRedux, [
      `${tag}::Response`,
      "error",
      "response"
    ])
  )

  // Choose whether to use typed form or raw text field to generate request body
  const whichFormData = simpleSelectorGet(
    props.simpleRedux,
    [`${tag}::RequestBodyFormInputType`, "data"],
    false
  )
    ? "RAW"
    : "FORM"
  const formData =
    whichFormData === "RAW"
      ? simpleSelectorGet(props.simpleRedux, [`${tag}::RawRequestBody`, "data"])
      : getFormData(action, props.simpleRedux, tag, props.webActionsRaw)

  // Open request section if method has a body
  if (
    methodHasBody(method) &&
    simpleSelectorGet(props.simpleRedux, [
      `${tag}::ButtonRequestBody`,
      "data"
    ]) == undefined
  ) {
    props.simpleMergeData(`${tag}::ButtonRequestBody`, true)
  }

  // Open request body form section if method has body
  if (
    methodHasBody(method) &&
    simpleSelectorGet(props.simpleRedux, [
      `${tag}::ButtonFormRequestBody`,
      "data"
    ]) == undefined
  ) {
    props.simpleMergeData(`${tag}::ButtonFormRequestBody`, true)
  }

  // Open the response section if request has been sent
  if (
    simpleSelectorGet(props.simpleRedux, [`${tag}::ButtonResponse`, "data"]) ==
      undefined &&
    responseData
  ) {
    props.simpleMergeData(`${tag}::ButtonResponse`, true)
  }
  return (
    <Collapse isOpen={isOpen}>
      <InputGroup
        defaultValue={action.pathPattern}
        onChange={onChangeFnCall(props.simpleMergeData, `${tag}::URL`)}
        placeholder={
          "Request URL: absolute ( http://your.url.com/to/send/a/request/to/ ) or internal service endpoint ( /service/web/action )"
        }
        type={"url"}
      />
      <FlexContainer>
        <div css={cssColumn}>
          <Menu css={cssMetadataMenu}>
            <RequestBodyForm
              action={action}
              method={method}
              tag={tag}
              {...props}
            />
          </Menu>
        </div>
        <div css={cssColumn}>
          <ControlGroup>
            <HTMLSelect
              large={true}
              onChange={onChangeFnCall(props.simpleMergeData, `${tag}::Method`)}
              options={action.httpMethod.sort()}
              value={method}
            />
            <Button
              css={cssButton}
              large={true}
              onClick={(event: any) => {
                props.simpleMergeData(`${tag}::ButtonRequestBody`, false)
                HTTPMethodDispatch(props)[method](
                  `${tag}::Response`,
                  url,
                  isOpen && formData
                )
              }}
              intent={HTTPMethodIntent[method]}
              loading={simpleSelectorGet(props.simpleRedux, [
                `${tag}::Response`,
                "loading"
              ])}
              text={"Submit"}
            />
          </ControlGroup>
          <Menu css={cssMetadataMenu}>
            {methodHasBody(method) ? (
              <MetadataCollapse
                label={`${url}`}
                tag={`${tag}::ButtonRequestBody`}
                text={"Request"}
              >
                <MetadataCopyToClipboard
                  data={formData}
                  description={"Request Body"}
                />
                <CodePreContainer>
                  {JSON.stringify(isOpen && formData, null, 2)}
                </CodePreContainer>
              </MetadataCollapse>
            ) : (
              <MetadataCollapse
                content={[]}
                label={`${url}`}
                tag={`${tag}::ButtonRequestBody`}
                text={"Request"}
              />
            )}
            <MetadataCollapse
              labelElement={
                <StatusTagComponent
                  status={simpleSelectorGet(
                    props.simpleRedux,
                    [`${props.tag}::Response`, "data", "status"],
                    simpleSelectorGet(
                      props.simpleRedux,
                      [`${props.tag}::Response`, "error", "status"],
                      0
                    )
                  )}
                  statusText={simpleSelectorGet(
                    props.simpleRedux,
                    [`${props.tag}::Response`, "data", "statusText"],
                    simpleSelectorGet(
                      props.simpleRedux,
                      [`${props.tag}::Response`, "error", "statusText"],
                      ""
                    )
                  )}
                />
              }
              tag={`${tag}::ButtonResponse`}
              text={"Response"}
            >
              <div>
                <MetadataCopyToClipboard
                  data={responseData}
                  description={"Raw Response"}
                />
                <CodePreContainer>
                  {JSON.stringify(responseData, null, 2)}
                </CodePreContainer>
                <MetadataCollapse
                  label={"Redux State"}
                  tag={`${tag}::ButtonRawResponse`}
                  text={"Raw Response"}
                >
                  <MetadataCopyToClipboard
                    data={response}
                    description={"Response"}
                  />
                  <CodePreContainer>
                    {JSON.stringify(response, null, 2)}
                  </CodePreContainer>
                </MetadataCollapse>
              </div>
            </MetadataCollapse>
          </Menu>
        </div>
      </FlexContainer>
    </Collapse>
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SendRequestCollapseContainer)
