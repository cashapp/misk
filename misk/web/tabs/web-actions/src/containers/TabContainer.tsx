import { H1 } from "@blueprintjs/core"
import * as React from "react"
import { connect } from "react-redux"
import { WebActionsContainer } from "../containers"
import {
  IDispatchProps,
  IState,
  mapDispatchToProps,
  mapStateToProps
} from "../ducks"

class TabContainer extends React.Component<IState & IDispatchProps, IState> {
  componentDidMount() {
    this.props.webActionsMetadata()
  }

  render() {
    return (
      <div>
        <H1>Web Actions</H1>
        <WebActionsContainer tag={"WebActions"} />
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(TabContainer)
