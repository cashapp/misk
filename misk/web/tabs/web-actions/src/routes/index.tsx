import * as React from "react"
import { Route, Switch } from "react-router"
import { TabContainer as OldTabContainer } from "../containers"
import TabContainer from "../rewrite/TabContainer"

const routes = (
  <div>
    <Switch>
      <Route path="/_admin/web-actions/" component={OldTabContainer} />
      <Route path="/_admin/web-actions-beta/" component={TabContainer} />
      {/* Do not include a Route without a path or it will display during on all tabs */}
    </Switch>
  </div>
)

export default routes
