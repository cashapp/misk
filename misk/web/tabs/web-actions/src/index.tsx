import { createApp, createIndex } from "@misk/core"
import * as Ducks from "./ducks"
import routes from "./routes"
export * from "./containers"

createIndex("web-actions", createApp(routes), Ducks)
