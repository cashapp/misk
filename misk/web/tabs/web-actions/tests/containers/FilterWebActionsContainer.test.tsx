import React from "react"
import { cleanup } from "@testing-library/react"
import { FilterWebActionsContainer } from "../../src/containers"
import { renderWithRedux } from "../upstreamableTestUtilities"

describe("FilterWebActionsContainer", () => {
  afterEach(cleanup)
  it("FilterWebActionsContainer can render with redux", () => {
    const { asFragment } = renderWithRedux(
      <FilterWebActionsContainer tag={"FilterTag"} />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
