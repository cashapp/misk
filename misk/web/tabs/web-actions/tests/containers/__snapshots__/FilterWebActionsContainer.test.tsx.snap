// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterWebActionsContainer FilterWebActionsContainer can render with redux 1`] = `
<DocumentFragment>
  <div
    class="bp3-control-group bp3-fill"
  >
    <div
      class="bp3-html-select bp3-large"
    >
      <select>
        <option
          value="All Metadata"
        >
          All Metadata
        </option>
        <option
          value="Allowed Capabilities"
        >
          Allowed Capabilities
        </option>
        <option
          value="Allowed Services"
        >
          Allowed Services
        </option>
        <option
          value="Application Interceptor"
        >
          Application Interceptor
        </option>
        <option
          value="HTTP Method"
        >
          HTTP Method
        </option>
        <option
          value="Function"
        >
          Function
        </option>
        <option
          value="Function Annotations"
        >
          Function Annotations
        </option>
        <option
          value="Name"
        >
          Name
        </option>
        <option
          value="Network Interceptor"
        >
          Network Interceptor
        </option>
        <option
          value="Parameter Types"
        >
          Parameter Types
        </option>
        <option
          value="Path Pattern"
        >
          Path Pattern
        </option>
        <option
          value="Request Type"
        >
          Request Type
        </option>
        <option
          value="Response Type"
        >
          Response Type
        </option>
      </select>
      <span
        class="bp3-icon bp3-icon-double-caret-vertical"
        icon="double-caret-vertical"
      >
        <svg
          data-icon="double-caret-vertical"
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <desc>
            double-caret-vertical
          </desc>
          <path
            d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      class="bp3-input-group bp3-large"
    >
      <input
        class="bp3-input"
        placeholder="Filter Web Actions"
        style="padding-right: 0px;"
        type="text"
      />
      <span
        class="bp3-input-action"
      >
        <span />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
